src/Components/ShopcartButton/ShopcartButton.js

```diff
@@ -0,0 +1,62 @@
+import styled from "styled-components";
+import ShoppingCartContext from "../../state/ShoppingCartContext";
+import { useContext } from "react";
+import replaceInArray from "../../utils/replaceInArray";
+
+const AddToShopcartButton = styled.button`
+  display: block;
+  margin-left: auto;
+  margin-right: auto;
+  padding: 5px 20px 5px 20px;
+  background-color: rgba(0, 150, 255, 0.8);
+  font-size: 20px;
+  font-weight: bold;
+  text-decoration: none;
+  color: black;
+  border-radius: 15px;
+
+  &:hover {
+    background-color: rgba(0, 150, 255, 0.2);
+  }
+
+  &:disabled {
+    background-color: rgba(175, 175, 175);
+  }
+`;
+
+const ShopcartButton = ({ product, quantity: addedQuantity }) => {
+  const addItemToShopcart = () => {
+    let shopcartRecordIdx = items.findIndex(
+      (item) => item.product.id === product.id
+    );
+
+    if (shopcartRecordIdx < 0) {
+      let newShopcartRecord = {
+        quantity: Math.min(product.data.stock, addedQuantity),
+        product: product,
+      };
+      setItems([...items, newShopcartRecord]);
+    } else {
+      let updatedItem = {
+        ...items[shopcartRecordIdx],
+        quantity: Math.min(
+          product.data.stock,
+          items[shopcartRecordIdx].quantity + addedQuantity
+        ),
+      };
+
+      setItems(replaceInArray(updatedItem, shopcartRecordIdx, items));
+    }
+  };
```

Comment:
Creo que todas estas líneas podrían ser solo una función en utils.js para validar la existencia de un producto en el state, donde te regrese los items nuevos ya computados, de manera que no tengas que hacer el `setItems` de cada uno dentro del `if` o el `else`, sino que simplemente mandas llamar esa función dentro del `setItems` en este componente.

Es decir, asumiendo que la función ya existe quedaría algo así.
```suggestion
  const addItemToShopcart = () => {
      setItems(getNewShopcar(items));
  };
```